### PR TITLE
chore: added memory leakage tests

### PR DESCRIPTION
**Pull request by Shahroz16**
Updated tags so they can be reused while testing with Appium as well.

### CODE REVIEW

**Review by mrehan27 — APPROVED** [review]
Looks good to me 👍🏻

Can having same tests for Java Layout will help us? If yes, can we reuse some of the code to run this for Java/or all sample apps?

File: sdk/src/main/java/io/customer/sdk/CustomerIO.kt [hunk 1]
```diff
@@ -100,7 +100,11 @@ class CustomerIO internal constructor(
         @InternalCustomerIOApi
         @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
         fun clearInstance() {
-            instance = null
+            instance?.let {
+                (it.diGraph.context as Application)
+                    .unregisterActivityLifecycleCallbacks(it.diGraph.activityLifecycleCallbacks)
+                instance = null
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply sdk/src/main/java/io/customer/sdk/CustomerIO.kt:106]: Good call, we can probably reuse that method, let me try.

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        // Clear the CustomerIO instance.
+        CustomerIO.clearInstance()
+
+        // Call the System.gc() method to force the garbage collector to run.
+        System.gc()
+
+        // Assert that the AppWatcher object does not have any retained objects.
+        assertEquals(AppWatcher.objectWatcher.hasRetainedObjects, false)
+        assertEquals(AppWatcher.objectWatcher.retainedObjectCount, 0)
+    }
```
Comment (levibostian) [inline samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:242]: There is a lot going on in the test function. That's not a bad thing, but I wonder for maintainability purposes. 

If I was to need to modify this test function in the future, what do I need to know about? 

Is the test function long to try and fill up memory? Is it long because it's trying to call many different functions of the SDK?

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
```
Comment (levibostian) [inline samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:38]: Since this is an instrumentation test, at this time, we do not run this test automatically on the CI. 

We can and I believe we should. Mind creating a ticket to prioritize this?

Comment (Shahroz16) [reply samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:38]: Good idea, here is the [PR](https://github.com/customerio/customerio-android/pull/233) to run instrumentation tests of sample apps as part of the workflow.

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        }
+
+        CustomerIO.Builder(
+            siteId = configuration.siteId,
+            apiKey = configuration.apiKey,
+            appContext = appContext as Application
+        ).apply {
+            setLogLevel(CioLogLevel.DEBUG)
+            setBackgroundQueueMinNumberOfTasks(5)
+            setBackgroundQueueSecondsDelay(5.0)
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:75]: No specific reason wanted to demonstrate it as an option.

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        // wait for 500ms
+        Thread.sleep(500)
+
+        // Click the random event button 50 times.
+        val randomEvent =
+            composeTestRule.onNodeWithTag(appContext.getString(R.string.acd_random_event_button))
+
+        for (i in 0..50) {
+            randomEvent.performClick()
+        }
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:127]: To test edge cases so we don't just run the best case, this would add 50 items to the queue and might as well take more memory.

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                        .setEventListener(InAppMessageEventListener()).build()
+                )
+            )
+            addCustomerIOModule(ModuleMessagingPushFCM())
+            build()
+        }
+
+        // Delete all users. So we always land in `Login Screen`
+        runBlocking {
+            userPreferenceRepository.deleteAllUsers()
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:88]: I can see the value in it.

File: samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt (hunk 2 above)
Comment (Shahroz16) [reply samples/kotlin_compose/src/androidTest/java/io/customer/android/sample/kotlin_compose/MemoryLeakageInstrumentedTest.kt:242]: Both, fill out memory and check each part of the app to make sure we know at what point the memory spike came and the possible culprit.

File: samples/kotlin_compose/src/main/java/io/customer/android/sample/kotlin_compose/data/persistance/AppDatabase.kt [hunk 7]
```diff
@@ -7,7 +7,7 @@ import androidx.room.RoomDatabase
 import io.customer.android.sample.kotlin_compose.R
 import io.customer.android.sample.kotlin_compose.data.models.User
 
-@Database(entities = [User::class], version = 1, exportSchema = true)
+@Database(entities = [User::class], version = 1, exportSchema = false)
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply samples/kotlin_compose/src/main/java/io/customer/android/sample/kotlin_compose/data/persistance/AppDatabase.kt:10]: `exportSchema` by default is `true` and keep on giving warning to provide a path for export. We dont need to export the schema as we are not documenting the sample app schema, hence false.

[the remark this answers is not in the document]
Comment (Shahroz16) [reply samples/kotlin_compose/src/main/java/io/customer/android/sample/kotlin_compose/data/persistance/AppDatabase.kt:10]: `exportSchema` by default is `true` and keep on giving warning to provide a path for export. We dont need to export the schema as we are not documenting the sample app schema, hence false.

**Review by Shahroz16 — COMMENTED** [review]
Added the refactor as part of https://github.com/customerio/customerio-android/pull/233
